### ISSUE TITLE
Duplicate output

### ISSUE THREAD

**Issue by MakotoE** [rust-lang/rust-clippy#7356]
`cargo clippy` is outputting duplicate warnings. As a minimal example, I created a new directory, ran `cargo init --lib`, and replaced `lib.rs` with:

```rust
pub fn f() {
    println!("{}", 0 + 0);
}
```

Running `cargo clippy` outputs

```
    Checking rs-test v0.1.0 (/home/makoto/Downloads/rs-test)
warning: the operation is ineffective. Consider reducing it to `0`
 --> src/lib.rs:2:20
  |
2 |     println!("{}", 0 + 0);
  |                    ^^^^^
  |
  = note: `#[warn(clippy::identity_op)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op

warning: the operation is ineffective. Consider reducing it to `0`
 --> src/lib.rs:2:20
  |
2 |     println!("{}", 0 + 0);
  |                    ^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op

warning: 2 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
```

I expected to see:
```
    Checking rs-test v0.1.0 (/home/makoto/Downloads/rs-test)
warning: the operation is ineffective. Consider reducing it to `0`
 --> src/lib.rs:2:20
  |
2 |     println!("{}", 0 + 0);
  |                    ^^^^^
  |
  = note: `#[warn(clippy::identity_op)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op

warning: 1 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
```

I found [this similar issue](https://github.com/rust-lang/rust-clippy/issues/1365) but it applies to binaries, whereas I found that the issue I am experiencing applies to both binaries and libraries.

### Meta

- `cargo clippy -V`: `clippy 0.1.54 (539d7bd 2021-06-14)`
- `rustc -Vv`:
```
rustc 1.55.0-nightly (539d7bd39 2021-06-14)
binary: rustc
commit-hash: 539d7bd3998d9bfed14c264eacda30097a4ea768
commit-date: 2021-06-14
host: x86_64-unknown-linux-gnu
release: 1.55.0-nightly
LLVM version: 12.0.1
```

**Comment by MakotoE** [comment]
This issue also happens with rust stable, 1.52.1.
```
rustc 1.52.1 (9bc8c42bb 2021-05-09)
binary: rustc
commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
commit-date: 2021-05-09
host: x86_64-unknown-linux-gnu
release: 1.52.1
LLVM version: 12.0.0
```

**Comment by MakotoE** [comment]
[Clippy on Rust playground does this as well](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=07b17e26fb2f61b5a44a4c23111593ec).

**Comment by MakotoE** [comment]
This seems to only happen with rule `identity_op`.

**Comment by HKalbasi** [comment]
This is because both of zeros are `identity_op`. So it will generate one warning for left zero and one for right zero. If you change it to `0 + 1` you see just one warning and if you change it to `0 + 0 + 0` you will see three warnings. It can be better but I think it doesn't matter in real world.

**Comment by MakotoE** [comment]
Thanks. I should have checked the linked page for the explanation.
